nav2_loopback_sim/src/loopback_simulator.cpp

```diff
@@ -0,0 +1,227 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <memory>
+
+#include "nav2_loopback_sim/loopback_simulator.hpp"
+
+namespace nav2_loopback_sim
+{
+
+LoopbackSimulator::LoopbackSimulator(const rclcpp::NodeOptions & options)
+: Node("loopback_simulator_node", options)
+{
+  // Initialize subscribers
+  vel_subscriber_ = this->create_subscription<geometry_msgs::msg::Twist>(
+    "cmd_vel", 10, std::bind(&LoopbackSimulator::twistCallback, this, std::placeholders::_1));
+
+  init_pose_subscriber_ = create_subscription<geometry_msgs::msg::PoseWithCovarianceStamped>(
+    "initialpose", 10,
+    std::bind(&LoopbackSimulator::initPoseCallback, this, std::placeholders::_1));
+
+  // Initialize publishers
+  scan_publisher_ = this->create_publisher<sensor_msgs::msg::LaserScan>("scan", 10);
+
+  // Frequency parameter for timer function
+  double timer_frequency{0.0};
+  declare_parameter("timer_frequency", 10.0);
+  get_parameter("timer_frequency", timer_frequency);
+
+  // Delta time parameter for forward integration
+  declare_parameter("dt", 0.1);
+  get_parameter("dt", dt_);
+
+  // Create a timer
+  timer_ = create_wall_timer(
+    std::chrono::duration<double>(1.0 / timer_frequency),
+    std::bind(&LoopbackSimulator::timerCallback, this));
+
+  // Transforms
+  tf_broadcaster_ = std::make_unique<tf2_ros::StaticTransformBroadcaster>(*this);
+  tf_buffer_ = std::make_unique<tf2_ros::Buffer>(this->get_clock());
+  tf_listener_ = std::make_shared<tf2_ros::TransformListener>(*tf_buffer_);
+}
+
+LoopbackSimulator::~LoopbackSimulator()
+{
+  RCLCPP_INFO(get_logger(), "Destroying %s", get_name());
+}
+
+
+void LoopbackSimulator::publishTransform(
+  const geometry_msgs::msg::Pose & pose,
+  const std::string & parent_frame,
+  const std::string & child_frame)
+{
+  geometry_msgs::msg::TransformStamped transformStamped;
+  transformStamped.header.stamp = this->now() + rclcpp::Duration::from_seconds(dt_);
+  transformStamped.header.frame_id = parent_frame;
+  transformStamped.child_frame_id = child_frame;
+  transformStamped.transform.translation.x = pose.position.x;
+  transformStamped.transform.translation.y = pose.position.y;
+  transformStamped.transform.translation.z = pose.position.z;
+  transformStamped.transform.rotation = pose.orientation;
+
+  tf_broadcaster_->sendTransform(transformStamped);
+}
+
+void LoopbackSimulator::twistCallback(const geometry_msgs::msg::Twist::SharedPtr msg)
+{
+  if (!init_pose_set_) {
+    RCLCPP_WARN(
+      rclcpp::get_logger("logger_name"),
+      "Initial pose has not been set. Ignoring incoming velocity messages.");
+    return;
+  }
+
+  // get base_link location in the odom frame
+  geometry_msgs::msg::TransformStamped transformStamped;
```

Comment:
If you just store the odom_to_base_link` transform in the class, you can just update it every time with new twist data. You don't need to look it up since you're the only publisher of that data. If you just store `new_position` as `odom_to_base_link_`, that can be used repeatedly and updated each time. It only just needs to be set to Identity on the first publication and then you're good

Reply:
Absolutely! That will be so much better